system/security/plugins/jwtSecurity/jwtCache.cpp

```diff
@@ -0,0 +1,410 @@
+}
+
+SecAccessFlags JWTUserInfo::_matchScopePerm(const PermissionMap& scopePermMap, const std::string& scope, SecAccessFlags defaultFlag) const
+{
+    // Ideally, this function should be traversing a tree of scope permissions, with each
+    // scope level a tree node; that would be much faster and would also avoid requiring
+    // ordering multiple patterns prior to calling _addScopePerm()
+    std::string         scopeAsExample(_convertPathname(_globToExample(scope)));
+    SecAccessFlags      accessValue = defaultFlag;
+    int                 longestScopeLen = 0;
+
+    for (PermissionMap::const_iterator x = scopePermMap.begin(); x != scopePermMap.end(); x++)
+    {
+        if (::fnmatch(x->first.c_str(), scopeAsExample.c_str(), (FNM_CASEFOLD | FNM_LEADING_DIR)) == 0)
+        {
+            // Use only the longest (most precise) scope
+            if (x->first.size() > longestScopeLen)
+            {
+                longestScopeLen = x->first.size();
+                accessValue = x->second;
+            }
+        }
+    }
+
+    return accessValue;
+}
+
+JWTUserInfo& JWTUserInfo::_eraseScopePerm(PermissionMap& scopePermMap, const std::string& scope)
+{
+    std::string         rewrittenScope(_convertPathname(scope));
+
+    scopePermMap.erase(rewrittenScope);
+
+    return *this;
+}
+
+JWTUserInfo& JWTUserInfo::_clearScopePerm(PermissionMap& scopePermMap)
+{
+    scopePermMap.clear();
+
+    return *this;
+}
+
+unsigned int JWTUserInfo::_countScopePerms(const PermissionMap& scopePermMap) const
+{
+    return scopePermMap.size();
+}
+
+//=============================================================================
+
+JWTUserCache::JWTUserCache()
+{
+}
+
+JWTUserCache::~JWTUserCache()
+{
+}
+
+bool JWTUserCache::has(const std::string& userName) const
+{
+    bool wasFound = false;
+
+    {
+        CriticalBlock block(crit);
+
+        wasFound = userPermMap.find(userName) != userPermMap.end();
+    }
+
+    return wasFound;
+}
+
+JWTUserCache& JWTUserCache::set(const std::string& userName, const JWTUserInfo& userInfo)
+{
+    {
+        CriticalBlock block(crit);
+
+        userPermMap[userName] = userInfo;
+    }
+
+    return *this;
+}
+
+bool JWTUserCache::get(const std::string& userName, JWTUserInfo& userInfo)
+{
+    UserPermissionMap::iterator foundIter;
+
+    {
+        CriticalBlock block(crit);
+
+        foundIter = userPermMap.find(userName);
+    }
+
+    if (foundIter != userPermMap.end())
+    {
+        userInfo = foundIter->second;
+        return true;
+    }
+
+    return false;
+}
+
+JWTUserCache& JWTUserCache::erase(const std::string& userName)
+{
+    {
+        CriticalBlock block(crit);
+
+        userPermMap.erase(userName);
+    }
+
+    return *this;
+}
+
+JWTUserCache& JWTUserCache::clear()
+{
+    {
+        CriticalBlock block(crit);
+
+        userPermMap.clear();
+    }
+
+    return *this;
+}
+
+unsigned int JWTUserCache::count() const
+{
+    return userPermMap.size();
```

Comment:
Should this map read also be guarded?

Reply:
std::map::size() is retrieved from a member variable.  While it is possible to get an incorrect value due to threaded access, a mutex does not add a guarantee that the value would be correct anyway.